content/blog/2025-05-13-bril2wasm.md

```diff
@@ -0,0 +1,55 @@
+
+Another challenge was working with code created by another group in the class, who were working toward the same deadline we were. Throughout the course of their Brilooped project, they naturally ran into adjustments to the Brilooped format, which then cascaded into affecting our project as well; as the two were developing concurrently, we did not have a well of extensive Briloop documentation to refer back to when developing our programs. However, we did have the advantage of being able to talk to Dev and Gerardo in real time. We could, and did, ask for help directly whenever we ran into trouble, and they were really really helpful in helping us get our input from Briloop working. We’re incredibly grateful for their help throughout this process.
+
+One more challenging aspect of the project was, for the limited capacity in which we used Gemini, wrangling the LLM to produce actually correct (or even executable) code; though it still probably saved us the time it would take to do the uninteresting bit ourselves, it was certainly not an effortless “get out of jail free” card.
+
+# Analysis
+
+## Testing Methodology
+We tested our code thoroughly on small examples as we implemented core functionality. This allowed us to catch errors before our code became too complex to fix. After completing our implementation, we tested on the Bril benchmarks core test suite. To do this, we first took Briloop core benchmark translations, provided by the Briloop group, and translated them into json files using their Briloop2json command. After, we ran the files through our code to output the WebAssembly translation, then ran this translation using Wasmtime. We compared the output of our WebAssembly files to the output from the original Bril files. With this method, we were able to confirm correctness, in the sense that the WebAssembly output program behaved identically on our test cases to the original Bril programs (1 test per benchmark).
+
+Note that our WebAssembly runtime prints all arguments one per line, instead of multiple arguments on the same line. Thus, our test cases ignored whitespace differences.
```

Comment:
This should be solvable. Check out the runtime libraries that I linked above: the idea is to compile `print a b c;` into something like:

```c
_bril_print_int(a);
_bril_print_sep();
_bril_print_int(b);
_bril_print_sep();
_bril_print_int(c);
_bril_print_end();
```